### ISSUE TITLE
[Bug]: Failed to clone Git submodule

### ISSUE THREAD

**Issue by marinofranz** [coollabsio/coolify#2731]
### Description

Hi,

I've just created an application on Coolify that pulls from a private GitHub repository. Said repository has a submodule, and Coolify doesn't seem to clone it correctly. I've made sure that **Submodules** are turned on under **Advanced** settings for the application. Is there something permission wise that needs to be addressed on the GitHub App itself, or is this a bug?

There are 2 other issues that report the same problem, which have been closed, but with no resolution:
#795 #1050 

Happy to provide more information if needed 🙂 

### Minimal Reproduction (if possible, example repository)

1. Create an application
2. Choose Dockerfile as Build Pack (I've also tried Nixpacks)
3. Deploy
4. An error occurs and is displayed in the deployment log

### Exception or Error

```
Cloning into '/artifacts/q88kow8'...
Submodule 'prisma' (https://github.com/???/prisma.git) registered for path 'prisma'
Cloning into '/artifacts/q88kow8/prisma'...
fatal: could not read Username for '[https://github.com':](https://github.com'/) No such device or address
fatal: clone of 'https://github.com/???/prisma.git' into submodule path '/artifacts/q88kow8/prisma' failed
Failed to clone 'prisma'. Retry scheduled
```

> _Organization name has been redacted from the logs._

### Version

v4.0.0-beta.306

**Comment by marinofranz** [comment]
Right, I've figured it out. For those who are having the same problem, you need to make sure that the URL for your submodules are relative, and not absolute. For some reason (something to do with authentication I suppose), the submodules will only clone when using the SSH protocol.

These are the changes I made to my `.gitmodules` file.
```diff
[submodule "prisma"]
	path = prisma
-	url = https://github.com/???/prisma.git
```
```diff
[submodule "prisma"]
	path = prisma
+	url = ../prisma.git
```

This _might_ be worth mentioning in the Coolify knowledge base under the [Applications page](https://coolify.io/docs/knowledge-base/applications#git-submodules).

**Comment by kyuumeitai** [comment]
Live saving! great job finding it @marinofranz

**Comment by musticode** [comment]
For coolify : 
- Add new git source and authenticate to github
- In project deployment configuration page, go to Advanced tab. Below of Git section, Submodules should be unchecked.

**Comment by marinofranz** [comment]
> For coolify :
> 
> * Add new git source and authenticate to github
> * In project deployment configuration page, go to Advanced tab. Below of Git section, Submodules should be unchecked.

I haven't tested this yet as my [posted solution](https://github.com/coollabsio/coolify/issues/2731#issuecomment-2199209294) works fine but disabling submodules to make them work seems a bit counter-intuitive..

**Comment by loudar** [comment]
Disabling under "Advanced" also worked for me.

Previously failed when they were enabled when trying to use private submodules from Github - despite having an authenticated GH App connected!

**Comment by justincavery** [comment]
I agree with @marinofranz, this is super unintuitive to turn off git submodules (although I tested this and it _does_ get past the same error which I was having.

**Comment by oribenez** [comment]
> Right, I've figured it out. For those who are having the same problem, you need to make sure that the URL for your submodules are relative, and not absolute. For some reason (something to do with authentication I suppose), the submodules will only clone when using the SSH protocol.
> 
> These are the changes I made to my `.gitmodules` file.
> 
> [submodule "prisma"]
> 	path = prisma
> -	url = https://github.com/???/prisma.git
> [submodule "prisma"]
> 	path = prisma
> +	url = ../prisma.git
> This _might_ be worth mentioning in the Coolify knowledge base under the [Applications page](https://coolify.io/docs/knowledge-base/applications#git-submodules).

I've tried this but not working.
Could you explain how to do it or give your dockerfile as an example?

**Comment by marwie** [comment]
Got the same issue here - the submodule that fails to pull is added to my coolify source github app with the same access rights as the root repository

The solution with changing the git module url didnt work for me, however turning `submodules` OFF in the Advanced tab *did* work but as others have already pointed out this is super unintuinitive and seems wrong.

Would be great to get a reaction from the coolify team on this

**Comment by sam-laister** [comment]
Same issue

**Comment by jonas-klesen** [comment]
I had a similar to issue as [closed issue 1050](https://github.com/coollabsio/coolify/issues/1050): cant clone a private repo with a private submodule when using the Git integration.
After trying everything else, the relative path also worked in this case. I did not need to uncheck the "submodules" checkbox.

**Comment by CharlesFr** [comment]
@jonas-klesen did you setup a seperate ssh or did you get it working with Git App Coolify integration? I've tried everything and nothing seems to work.

**Comment by jonas-klesen** [comment]
Can't check right now, but I think I added the server ssh key(s) to the
github accounts ssh keys. That was it iirc.

**Comment by CharlesFr** [comment]
got it working with relative paths, I think the main limitation there is that all submodules must come from the same GitHub Org/User

**Comment by sayedhfatimi** [comment]
> Right, I've figured it out. For those who are having the same problem, you need to make sure that the URL for your submodules are relative, and not absolute. For some reason (something to do with authentication I suppose), the submodules will only clone when using the SSH protocol.
> 
> These are the changes I made to my `.gitmodules` file.
> 
> [submodule "prisma"]
> 	path = prisma
> -	url = https://github.com/???/prisma.git
> [submodule "prisma"]
> 	path = prisma
> +	url = ../prisma.git
> This _might_ be worth mentioning in the Coolify knowledge base under the [Applications page](https://coolify.io/docs/knowledge-base/applications#git-submodules).

had same problem this morning, this worked for me too, thanks
